### ISSUE TITLE
CCD not restarted anymore, Kubernetes minor updates not working anymore

### ISSUE THREAD

**Issue by Gerrit91** [metal-stack/os-metal-extension#12]
I guess we have a regression bug in this controller. The cloud-config-downloader does not run anymore on our nodes and thus, changes from the OSC are not being reflected (e.g. Kubelet updates).

Seems to be related to: https://github.com/gardener/gardener/pull/3449

**Comment by mwennrich** [comment]
```
kubelet.service - kubelet daemon
     Loaded: loaded (/etc/systemd/system/kubelet.service; enabled; vendor preset: enabled)
     Active: active (running) since Wed 2021-04-21 20:39:20 CEST; 5 days ago
       Docs: https://kubernetes.io/docs/admin/kubelet
    Process: 312466 ExecStartPre=/usr/bin/env sh -c ID="$(/usr/bin/docker run --rm -d -v /opt/bin:/opt/bin:rw eu.gcr.io/gardener-project/hyperkube:v1.19.10)"; /usr/bin/docker cp "$ID":/kubelet /opt/bin; docker stop "$ID"; chmod +x /opt/bin/kubelet (code=exited, status=0/SUCCESS)
   Main PID: 312840 (kubelet)
      Tasks: 60 (limit: 114351)
     Memory: 190.9M
     CGroup: /system.slice/kubelet.service
             └─312840 /opt/bin/kubelet --bootstrap-kubeconfig=/var/lib/kubelet/kubeconfig-bootstrap --config=/var/lib/kubelet/config/kubelet --cni-bin-dir=/opt/cni/bin/ --cni-conf-dir=/etc/cni/net.d/ --image-pull-progress-deadline=1m --kubeconfig=/var/lib/kubelet/kubeconfig-real --network-plugin=cni --v=2 --container-runtime=remote --container-runtime-endpoint=unix:///run/containerd/containerd.sock --cloud-provider=external
```

ccd did run, ccd preloaded the new hyperkube images, kubelet.service has been modified, kubelet has been restarted. But /opt/bin/kubelet has not been modified.

Apparently, ExecStartPre did not run correctly.

**Comment by mwennrich** [comment]
```
Apr 28 13:29:59 shoot--pz9cjf--chngtest-group-0-659b8-dp66h download-cloud-config.sh[20647]: Checking whether to preload hyperkube from eu.gcr.io/gardener-project/hyperkube:v1.19.10
Apr 28 13:29:59 shoot--pz9cjf--chngtest-group-0-659b8-dp66h download-cloud-config.sh[20647]: Preloading hyperkube from eu.gcr.io/gardener-project/hyperkube:v1.19.10
Apr 28 13:30:01 shoot--pz9cjf--chngtest-group-0-659b8-dp66h download-cloud-config.sh[20666]: v1.19.10: Pulling from gardener-project/hyperkube
Apr 28 13:30:07 shoot--pz9cjf--chngtest-group-0-659b8-dp66h download-cloud-config.sh[20666]: Status: Downloaded newer image for eu.gcr.io/gardener-project/hyperkube:v1.19.10
Apr 28 13:30:07 shoot--pz9cjf--chngtest-group-0-659b8-dp66h download-cloud-config.sh[20666]: eu.gcr.io/gardener-project/hyperkube:v1.19.10
Apr 28 13:30:07 shoot--pz9cjf--chngtest-group-0-659b8-dp66h download-cloud-config.sh[20647]: Seen newer cloud config version
Apr 28 13:30:07 shoot--pz9cjf--chngtest-group-0-659b8-dp66h download-cloud-config.sh[20647]: Successfully applied new cloud config version

Apr 28 13:30:08 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Stopping kubelet daemon...
Apr 28 13:30:08 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: kubelet.service: Succeeded.
Apr 28 13:30:08 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Stopped kubelet daemon.
Apr 28 13:30:08 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Starting kubelet daemon...
Apr 28 13:30:08 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Stopping Kubelet-monitor daemon...
Apr 28 13:30:08 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: kubelet-monitor.service: Succeeded.
Apr 28 13:30:08 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Stopped Kubelet-monitor daemon.
Apr 28 13:30:08 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Starting Kubelet-monitor daemon...

Apr 28 13:30:09 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Stopping Docker Application Container Engine...
Apr 28 13:30:09 shoot--pz9cjf--chngtest-group-0-659b8-dp66h dockerd[3048]: time="2021-04-28T13:30:09.448778820+02:00" level=info msg="Daemon shutdown complete"
Apr 28 13:30:09 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: cloud-config-downloader.service: Succeeded.
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h env[21163]: docker: error during connect: Post http://%2Fvar%2Frun%2Fdocker.sock/v1.41/containers/create: EOF.
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h env[21163]: See 'docker run --help'.
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h env[21166]: docker: error during connect: Post http://%2Fvar%2Frun%2Fdocker.sock/v1.41/containers/create: EOF.
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h env[21166]: See 'docker run --help'.
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: docker.service: Succeeded.
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Stopped Docker Application Container Engine.
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Stopping containerd container runtime...
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: containerd.service: Succeeded.
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Stopped containerd container runtime.
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: containerd-initializer.service: Succeeded.
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Stopped Containerd initializer.
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Stopping Containerd initializer...
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Starting Containerd initializer...
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Finished Containerd initializer.

Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Starting containerd container runtime...
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Started containerd container runtime.
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Starting Docker Application Container Engine...
Apr 28 13:30:10 shoot--pz9cjf--chngtest-group-0-659b8-dp66h dockerd[21331]: time="2021-04-28T13:30:10.639726358+02:00" level=info msg="Starting up"
Apr 28 13:30:11 shoot--pz9cjf--chngtest-group-0-659b8-dp66h dockerd[21331]: time="2021-04-28T13:30:11.987918820+02:00" level=info msg="Daemon has completed initialization"
Apr 28 13:30:12 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Started Docker Application Container Engine.
Apr 28 13:30:12 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Started Kubelet-monitor daemon.
Apr 28 13:30:12 shoot--pz9cjf--chngtest-group-0-659b8-dp66h systemd[1]: Started kubelet daemon.
```

Seems the `ExecStartPre` `docker run` and `docker cp` are executed while docker.service is still stopped.

And docker is restarted because of this:
```
    systemctl enable updatecacerts.service && systemctl restart --no-block updatecacerts.service
```
```
[Unit]
Description=Update CA bundle at /etc/ssl/certs/ca-certificates.crt
# Since other services depend on the certificate store run this early
DefaultDependencies=no
Wants=systemd-tmpfiles-setup.service clean-ca-certificates.service
After=systemd-tmpfiles-setup.service clean-ca-certificates.service
Before=sysinit.target kubelet.service
ConditionPathIsReadWrite=/etc/ssl/certs
ConditionPathExists=!/var/lib/kubelet/kubeconfig-real
[Service]
Type=oneshot
ExecStart=/usr/sbin/update-ca-certificates --fresh
ExecStartPost=/bin/systemctl restart docker
[Install]
WantedBy=multi-user.target
```

**Comment by mwennrich** [comment]
https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/kubelet/component.go#L183-L191

**Comment by majst01** [comment]
> https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/kubelet/component.go#L183-L191

WTF ??

**Comment by mwennrich** [comment]
Seems to have been fixed with update to gardener 1.19.x.
